Sources/Hummingbird/Server/Request.swift

```diff
@@ -57,6 +63,150 @@ extension Request {
+    /// the date in the `If-Modified-Since` request header.
+    ///
+    /// The most common use case is to update a cached entity that has no associated ETag.
+    ///
+    /// - Parameters:
+    ///   - headers: Headers to add to response when condition fails
+    ///   - modificationDate: Modification date for this responses content
+    ///   - context: Request context
+    ///   - process: Closure to run if eTag is not in the `If-None-Match` header
+    /// - Returns: Response
+    public func ifModifiedSince(
+        headers: HTTPFields = [:],
+        modificationDate: Date,
+        context: some RequestContext,
+        process: () async throws -> ResponseGenerator
+    ) async throws -> Response {
+        var headers = headers
+        headers[.lastModified] = modificationDate.httpHeader
+        // `If-Modified-Since` headers are only applied to GET or HEAD requests
+        if self.method == .get || self.method == .head {
+            if let ifModifiedSinceHeader = self.headers[.ifModifiedSince] {
+                if let ifModifiedSinceDate = Date(httpHeader: ifModifiedSinceHeader) {
+                    // round modification date of file down to seconds for comparison
+                    let modificationDateTimeInterval = modificationDate.timeIntervalSince1970.rounded(.down)
+                    let ifModifiedSinceDateTimeInterval = ifModifiedSinceDate.timeIntervalSince1970
+                    if modificationDateTimeInterval <= ifModifiedSinceDateTimeInterval {
+                        return Response(status: .notModified, headers: headers)
+                    }
+                }
+            }
+        }
+        var response = try await process().response(from: self, context: context)
+        response.headers.append(contentsOf: headers)
+        return response
+    }
+
+    /// Conditional request which will only be processed if the modification date supplied is before
+    /// the date in the `If-Unmodified-Since` request header.
+    ///
+    /// In conjunction with non-safe methods like POST, this header can be used to implement an
+    /// optimistic concurrency control, as is done by some wikis: revision are rejected if the
+    /// stored document has been modified since the original was retrieved, avoiding conflicts.
+    ///
+    /// - Parameters:
+    ///   - headers: Headers to add to response when condition fails
+    ///   - modificationDate: Modification date for this responses content
+    ///   - context: Request context
+    ///   - process: Closure to run if eTag is not in the `If-None-Match` header
+    /// - Returns: Response
+    public func ifUnmodifiedSince(
+        headers: HTTPFields = [:],
+        modificationDate: Date,
+        context: some RequestContext,
+        process: () async throws -> ResponseGenerator
+    ) async throws -> Response {
+        var headers = headers
+        headers[.lastModified] = modificationDate.httpHeader
```

Comment:
Aren't `If-Unmodified-Since` and `If-Match` used for making changes, creating new eTag/modificationDate, making the current one obsolete?

Reply:
Main uses are to avoid issues, where server side content changes between two requests. eg User read wikis, sends update to that wiki. A 412 is returned is wiki was updated between read and updating to avoid the losing the update. The other case I know of is user is downloading file using ranges. They need to be sure they are downloading from the same file if they send a second request with a different range.